File: app-android/src/main/kotlin/org/kopiaKt/app/ui/screens/filebrowser/FileBrowserScreen.kt
package org.kopiaKt.app.ui.screens.filebrowser

import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.automirrored.filled.InsertDriveFile
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import org.kopiaKt.app.domain.model.FileEntry
import org.kopiaKt.app.domain.model.FileEntryType
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FileBrowserScreen(
    snapshotId: String,
    initialPath: String,
    onNavigateToPath: (String) -> Unit,
    onRestore: (String) -> Unit,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
    viewModel: FileBrowserViewModel = hiltViewModel()
) {
    val uiState by viewModel.uiState.collectAsState()

    LaunchedEffect(initialPath) {
        viewModel.loadDirectory(initialPath)
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = {
                    Text(
                        text = if (initialPath.isEmpty()) "/" else initialPath.split("/").lastOrNull() ?: "/",
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis
                    )
                },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, "Back")
                    }
                },
                actions = {
                    IconButton(
                        onClick = { onRestore(initialPath) },
                        modifier = Modifier.testTag("restore_button")
                    ) {
                        Icon(Icons.Default.Download, "Restore this folder")
                    }
                }
            )
        }
    ) { padding ->
        Column(
            modifier = modifier
                .fillMaxSize()
                .padding(padding)
        ) {
            // Breadcrumb navigation
            BreadcrumbRow(
                pathHistory = uiState.pathHistory,
                onNavigate = { path ->
                    if (path != uiState.currentPath) {
                        onNavigateToPath(path)
                    }
                }
            )

            HorizontalDivider()

            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .weight(1f)
            ) {
                when {
                    uiState.isLoading -> {
                        CircularProgressIndicator(
                            modifier = Modifier.align(Alignment.Center)
                        )
                    }
                    uiState.error != null -> {
                        Column(
                            modifier = Modifier
                                .align(Alignment.Center)
                                .padding(16.dp),
                            horizontalAlignment = Alignment.CenterHorizontally
                        ) {
                            Icon(
                                Icons.Default.Warning,
                                contentDescription = null,
                                tint = MaterialTheme.colorScheme.error,
                                modifier = Modifier.size(48.dp)
                            )
                            Spacer(modifier = Modifier.height(16.dp))
                            Text(
                                text = uiState.error ?: "Unknown error",
                                style = MaterialTheme.typography.bodyLarge,
                                color = MaterialTheme.colorScheme.error
                            )
                            Spacer(modifier = Modifier.height(16.dp))
                            Button(onClick = { viewModel.loadDirectory(initialPath) }) {
                                Text("Retry")
                            }
                        }
                    }
                    uiState.entries.isEmpty() -> {
                        Text(
                            text = "Empty folder",
                            modifier = Modifier
                                .align(Alignment.Center)
                                .padding(16.dp),
                            style = MaterialTheme.typography.bodyLarge,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                    else -> {
                        LazyColumn(
                            modifier = Modifier.fillMaxSize()
                        ) {
                            items(uiState.entries, key = { it.name }) { entry ->
                                FileEntryItem(
                                    entry = entry,
                                    onClick = {
                                        when (entry.type) {
                                            FileEntryType.DIRECTORY -> {
                                                val newPath = if (uiState.currentPath.isEmpty()) {
                                                    entry.name
                                                } else {
                                                    "${uiState.currentPath}/${entry.name}"
                                                }
                                                onNavigateToPath(newPath)
                                            }
                                            else -> {
                                                // Could show file details or restore single file
                                            }
                                        }
                                    }
                                )
                            }
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun BreadcrumbRow(
    pathHistory: List<PathSegment>,
    onNavigate: (String) -> Unit,
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState())
            .padding(horizontal = 16.dp, vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        pathHistory.forEachIndexed { index, segment ->
            if (index > 0) {
                Icon(
                    Icons.Default.ChevronRight,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.size(20.dp)
                )
            }
            TextButton(
                onClick = { onNavigate(segment.fullPath) },
                contentPadding = PaddingValues(horizontal = 8.dp, vertical = 4.dp)
            ) {
                Text(
                    text = segment.name,
                    style = MaterialTheme.typography.bodyMedium,
                    color = if (index == pathHistory.lastIndex) {
                        MaterialTheme.colorScheme.primary
                    } else {
                        MaterialTheme.colorScheme.onSurfaceVariant
                    }
                )
            }
        }
    }
}

@Composable
private fun FileEntryItem(
    entry: FileEntry,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    ListItem(
        modifier = modifier
            .testTag("file_entry_${entry.name}")
            .clickable(onClick = onClick),
        leadingContent = {
            Icon(
                imageVector = getIconForEntry(entry),
                contentDescription = null,
                tint = getIconColor(entry)
            )
        },
        headlineContent = {
            Text(
                text = entry.name,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
        },
        supportingContent = {
            Row {
                if (entry.type == FileEntryType.FILE) {
                    Text(
                        text = formatSize(entry.size),
                        style = MaterialTheme.typography.bodySmall
                    )
                    Spacer(modifier = Modifier.width(16.dp))
                }
                entry.modTime?.let { time ->
                    Text(
                        text = formatDateTime(time),
                        style = MaterialTheme.typography.bodySmall
                    )
                }
            }
        },
        trailingContent = {
            if (entry.type == FileEntryType.DIRECTORY) {
                Icon(
                    Icons.Default.ChevronRight,
                    contentDescription = "Navigate",
                    tint = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
    )
}

@Composable
private fun getIconForEntry(entry: FileEntry): ImageVector {
    return when (entry.type) {
        FileEntryType.DIRECTORY -> Icons.Default.Folder
        FileEntryType.FILE -> Icons.AutoMirrored.Filled.InsertDriveFile
        FileEntryType.SYMLINK -> Icons.Default.Link
        FileEntryType.UNKNOWN -> Icons.Default.HelpOutline
    }
}

@Composable
private fun getIconColor(entry: FileEntry): androidx.compose.ui.graphics.Color {
    return when (entry.type) {
        FileEntryType.DIRECTORY -> MaterialTheme.colorScheme.primary
        FileEntryType.SYMLINK -> MaterialTheme.colorScheme.tertiary
        else -> MaterialTheme.colorScheme.onSurfaceVariant
    }
}

private fun formatDateTime(instant: java.time.Instant): String {
    val formatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.SHORT)
        .withZone(ZoneId.systemDefault())
    return formatter.format(instant)
}

private fun formatSize(bytes: Long): String {
    return when {
        bytes < 1024 -> "$bytes B"
        bytes < 1024 * 1024 -> "${bytes / 1024} KB"
        bytes < 1024 * 1024 * 1024 -> "${bytes / (1024 * 1024)} MB"
        else -> String.format("%.1f GB", bytes / (1024.0 * 1024 * 1024))
    }
}

File: flutter_ui/lib/screens/file_browser_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import 'package:intl/intl.dart';
import '../bridge/kopia_bridge.g.dart';
import '../providers/file_browser_provider.dart';

/// Screen for browsing files in a snapshot.
class FileBrowserScreen extends ConsumerStatefulWidget {
  const FileBrowserScreen({
    super.key,
    required this.snapshotId,
    this.initialPath = '',
  });

  final String snapshotId;
  final String initialPath;

  @override
  ConsumerState<FileBrowserScreen> createState() => _FileBrowserScreenState();
}

class _FileBrowserScreenState extends ConsumerState<FileBrowserScreen> {
  late final ScrollController _scrollController;

  @override
  void initState() {
    super.initState();
    _scrollController = ScrollController();
    _scrollController.addListener(_onScroll);

    // Load initial directory
    WidgetsBinding.instance.addPostFrameCallback((_) {
      ref
          .read(fileBrowserProvider(widget.snapshotId).notifier)
          .loadDirectory(widget.initialPath);
    });
  }

  @override
  void dispose() {
    _scrollController.dispose();
    super.dispose();
  }

  void _onScroll() {
    if (_scrollController.position.pixels >=
        _scrollController.position.maxScrollExtent - 200) {
      ref.read(fileBrowserProvider(widget.snapshotId).notifier).loadMore();
    }
  }

  @override
  Widget build(BuildContext context) {
    final state = ref.watch(fileBrowserProvider(widget.snapshotId));

    final title = widget.initialPath.isEmpty
        ? '/'
        : widget.initialPath.split('/').last;

    return Scaffold(
      appBar: AppBar(
        title: Text(
          title,
          maxLines: 1,
          overflow: TextOverflow.ellipsis,
        ),
        leading: IconButton(
          icon: const Icon(Icons.arrow_back),
          onPressed: () => context.pop(),
        ),
        actions: [
          IconButton(
            icon: const Icon(Icons.download),
            onPressed: () => context.push(
              '/restore/${widget.snapshotId}',
              extra: widget.initialPath,
            ),
            tooltip: 'Restore this folder',
          ),
        ],
      ),
      body: Column(
        children: [
          // Breadcrumb navigation
          _BreadcrumbRow(
            pathHistory: state.pathHistory,
            onNavigate: (path) {
              if (path != state.currentPath) {
                context.go('/files/${widget.snapshotId}?path=$path');
              }
            },
          ),
          const Divider(height: 1),

          // Content
          Expanded(child: _buildContent(context, state)),
        ],
      ),
    );
  }

  Widget _buildContent(BuildContext context, FileBrowserState state) {
    // Loading state (initial)
    if (state.isLoading && state.entries.isEmpty) {
      return Semantics(
        identifier: 'file_browser_loading',
        child: const Center(child: CircularProgressIndicator()),
      );
    }

    // Error state
    if (state.error != null && state.entries.isEmpty) {
      return Center(
        child: Padding(
          padding: const EdgeInsets.all(16.0),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Icon(
                Icons.warning,
                size: 48,
                color: Theme.of(context).colorScheme.error,
              ),
              const SizedBox(height: 16),
              Text(
                state.error!,
                style: Theme.of(context).textTheme.bodyLarge?.copyWith(
                      color: Theme.of(context).colorScheme.error,
                    ),
                textAlign: TextAlign.center,
              ),
              const SizedBox(height: 16),
              FilledButton(
                onPressed: () => ref
                    .read(fileBrowserProvider(widget.snapshotId).notifier)
                    .loadDirectory(widget.initialPath),
                child: const Text('Retry'),
              ),
            ],
          ),
        ),
      );
    }

    // Empty state
    if (state.entries.isEmpty) {
      return Center(
        child: Text(
          'Empty folder',
          style: Theme.of(context).textTheme.bodyLarge?.copyWith(
                color: Theme.of(context).colorScheme.onSurfaceVariant,
              ),
        ),
      );
    }

    // List of entries
    return Semantics(
      identifier: 'file_browser_ready',
      child: ListView.builder(
        controller: _scrollController,
        itemCount: state.entries.length + (state.hasMore ? 1 : 0),
        itemBuilder: (context, index) {
          if (index == state.entries.length) {
            return const Center(
              child: Padding(
                padding: EdgeInsets.all(16.0),
                child: CircularProgressIndicator(),
              ),
            );
          }

          final entry = state.entries[index];
          return _FileEntryItem(
            entry: entry,
            onTap: () {
              if (entry.type == FileEntryType.directory) {
                final newPath = state.currentPath.isEmpty
                    ? entry.name
                    : '${state.currentPath}/${entry.name}';
                context.go('/files/${widget.snapshotId}?path=$newPath');
              }
            },
          );
        },
      ),
    );
  }
}

/// Breadcrumb navigation row.
class _BreadcrumbRow extends StatelessWidget {
  const _BreadcrumbRow({
    required this.pathHistory,
    required this.onNavigate,
  });

  final List<PathSegment> pathHistory;
  final void Function(String path) onNavigate;

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
      child: Row(
        children: [
          for (var i = 0; i < pathHistory.length; i++) ...[
            if (i > 0)
              Icon(
                Icons.chevron_right,
                size: 20,
                color: Theme.of(context).colorScheme.onSurfaceVariant,
              ),
            TextButton(
              onPressed: () => onNavigate(pathHistory[i].fullPath),
              style: TextButton.styleFrom(
                padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 4),
                minimumSize: Size.zero,
              ),
              child: Text(
                pathHistory[i].name,
                style: TextStyle(
                  color: i == pathHistory.length - 1
                      ? Theme.of(context).colorScheme.primary
                      : Theme.of(context).colorScheme.onSurfaceVariant,
                ),
              ),
            ),
          ],
        ],
      ),
    );
  }
}

/// List item for a file or directory entry.
class _FileEntryItem extends StatelessWidget {
  const _FileEntryItem({
    required this.entry,
    required this.onTap,
  });

  final FileEntry entry;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    return Semantics(
      identifier: 'file_entry_${entry.name}',
      label: entry.name,  // For screen readers
      child: ListTile(
        leading: Icon(
          _getIcon(),
          color: _getIconColor(context),
        ),
        title: Text(
          entry.name,
          maxLines: 1,
          overflow: TextOverflow.ellipsis,
        ),
        subtitle: Row(
          children: [
            if (entry.type == FileEntryType.file) ...[
              Text(
                _formatSize(entry.size),
                style: Theme.of(context).textTheme.bodySmall,
              ),
              const SizedBox(width: 16),
            ],
            if (entry.modTimeEpochMs != null)
              Text(
                _formatDateTime(entry.modTimeEpochMs!),
                style: Theme.of(context).textTheme.bodySmall,
              ),
          ],
        ),
        trailing: entry.type == FileEntryType.directory
            ? Icon(
                Icons.chevron_right,
                color: Theme.of(context).colorScheme.onSurfaceVariant,
              )
            : null,
        onTap: onTap,
      ),
    );
  }

  IconData _getIcon() {
    switch (entry.type) {
      case FileEntryType.directory:
        return Icons.folder;
      case FileEntryType.file:
        return Icons.insert_drive_file;
      case FileEntryType.symlink:
        return Icons.link;
      case FileEntryType.unknown:
        return Icons.help_outline;
    }
  }

  Color _getIconColor(BuildContext context) {
    switch (entry.type) {
      case FileEntryType.directory:
        return Theme.of(context).colorScheme.primary;
      case FileEntryType.symlink:
        return Theme.of(context).colorScheme.tertiary;
      default:
        return Theme.of(context).colorScheme.onSurfaceVariant;
    }
  }

  String _formatDateTime(int epochMs) {
    final dateTime = DateTime.fromMillisecondsSinceEpoch(epochMs);
    final formatter = DateFormat.yMd().add_jm();
    return formatter.format(dateTime);
  }

  String _formatSize(int bytes) {
    if (bytes < 1024) return '$bytes B';
    if (bytes < 1024 * 1024) return '${bytes ~/ 1024} KB';
    if (bytes < 1024 * 1024 * 1024) return '${bytes ~/ (1024 * 1024)} MB';
    return '${(bytes / (1024 * 1024 * 1024)).toStringAsFixed(1)} GB';
  }
}
